IMAGING-134: Fix JPEG handling for files encoded with RST markers

mgs255 opened the pull request:
Shamelessly pulled from @mgmechanics pull request for the same bug.

kinow commented:
PR related: #13

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -71,11 +73,23 @@ public boolean beginSOS() {
             sosSegment = new SosSegment(marker, sosSegmentBytes);
 
+            // read the payload of the scan, this is the remainder of image data after the header
+            // the payload contains the entropy-encoded segments (or ECS) divded by RST markers
```
s/divded/divided

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegInputStream.java:
```diff
@@ -15,31 +15,39 @@
-    JpegInputStream(final InputStream is) {
-        this.is = is;
+    JpegInputStream(final int[] interval) {
+        this.interval = interval;
```
`this.interval = Arrays.copyOf(interval, interval.length);` to prevent mutability issues? Not sure if there is any risk (i.e. if internal only, probably not at risk, but may be still a good idea to play defensively here)

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegInputStream.java:
```diff
@@ -57,4 +65,19 @@ public int nextBit() throws IOException, ImageReadException {
+     * @throws IllegalStateException if the stream hasn't any other value.
+     */
+    int read() {
+        if (this.hasNext() == false) {
```
`if (!this.hasNext())`

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+            }
+
+            // true if one of the RST markers was found
+            if ( foundFF == true && foundD0toD7 == true ) {
```
`if (foundFF && foundD0toD7)`

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+        boolean foundD0toD7 = false;
+        int pos = 0;
+        while ( pos < scanPayload.length ) {
+            if ( foundFF == true ) {
```
`if ( foundFF ) {`

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+            if ( foundFF == true && foundD0toD7 == true ) {
+                // we need to add the position after the current position because
+                // we had already read 0xFF and are now at 0xDn
+                intervalStarts.add( pos + 1 );
```
`intervalStarts.add(pos + 1);`, spaces

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+                }
+            }
+
+            if ( scanPayload[ pos ] == 0xFF ) {
```
`if (scanPayload[ pos ] == 0xFF) {`, spaces

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+        boolean foundFF = false;
+        boolean foundD0toD7 = false;
+        int pos = 0;
+        while ( pos < scanPayload.length ) {
```
`while (pos < scanPayload.length) {`, spaces

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+     */
+    static List<Integer> getIntervalStartPositions(final int[] scanPayload) {
+        final List<Integer> intervalStarts = new ArrayList<Integer>();
+        intervalStarts.add( 0 );
```
`intervalStarts.add(0);`, spaces

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -393,6 +418,74 @@ private void readMCU(final JpegInputStream is, final int[] preds, final Block[]
+     * @param scanPayload array to examine
+     * @return the start positions, the number of start positions is also the count of intervals
+     * while the number of restart markers found = number of start positions - 1
+     * because restart markers are between intervals
```
What about moving the longer text to the main body of the javadocs comment? i.e.

```java
     * Returns the positions of where each interval in the provided array starts. The number
     * of start positions is also the count of intervals while the number of restart markers
     * found is equal to the number of start positions minus one (because restart markers
     * are between intervals).
     * @param scanPayload array to examine
     * @return the start positions
```

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -122,8 +134,21 @@ public void visitSOS(final int marker, final byte[] markerBytes, final byte[] im
                 for (int x1 = 0; x1 < hSize * xMCUs; x1 += hSize) {
+                    // provide a the next interval if an interval is read until it's end
+                    // as long there are unread intervals available
+                    if (bitInputStream.hasNext() == false) {
```
`if (!bitInputStream.hasNext()) {`

kinow commented on src/main/java/org/apache/commons/imaging/formats/jpeg/decoder/JpegDecoder.java:
```diff
@@ -122,8 +134,21 @@ public void visitSOS(final int marker, final byte[] markerBytes, final byte[] im
             for (int y1 = 0; y1 < vSize * yMCUs; y1 += vSize) {
                 for (int x1 = 0; x1 < hSize * xMCUs; x1 += hSize) {
+                    // provide a the next interval if an interval is read until it's end
+                    // as long there are unread intervals available
```
Maybe

```
provide the next interval if an interval is read until its end
as long as there are unread intervals available
```
?

kinow commented:
The rest of the code looks good to me. Once the current items are addressed, LGTM :+1:

mgs255 commented:
Thanks for the thorough review.  Our code style at works adds lots of spaces.  I think I've addressed all of the points you raised.  

I also wrote a little tool to iterate over some customer image data we store in S3.  It loaded around 40GB of mainly jpg files when I left it running for some time.  So far I've not seen any problems.

kinow commented:
That was quick! Lolks good to me, merging once I arrive at work. And thanks for the feedback about your use case at your work. We might need to release alpha2 soon to get the new code tested too.
